### PR TITLE
Strip Bearer From Token

### PR DESCRIPTION
**Pull request by dpbackes**
- Some libraries use the "Bearer" prefix when doing token
  authentication. If we encounter a token with a "Bearer"
  prefix then just strip it off and proceed as normal.

### CODE REVIEW

**Comment by dpbackes** [comment]
The C# library I'm using does this and it'll be easier to deal with it here rather than doing the stripping in all our services. cc @tylerodonnell @eesmith from our discussion earlier today.

File: index.js [hunk 1]
```diff
@@ -26,6 +26,9 @@ const chooseKey = key =>
 
 const decode = partialRight(jwt.decode, [{ complete: true }])
 
+const stripBearer = token => 
+  token ? token.replace(/^Bearer /, '') : null
```
Comment (tylerodonnell) [inline index.js]: I think we should also strip the lowercase version. Thoughts?

**Comment by tylerodonnell** [comment]
So most of our services are already stripping `bearer` from the token - but I like having the option of Authentic stripping "Bearer" if it exists.

File: index.js (hunk 1 above)
Comment (dpbackes) [reply index.js]: don't see any reason not to

**Comment by flintinatux** [comment]
👀

File: index.js [hunk 2]
```diff
@@ -26,6 +26,9 @@ const chooseKey = key =>
 
 const decode = partialRight(jwt.decode, [{ complete: true }])
 
+const stripBearer = token => 
+  token ? token.replace(/^[B|b]earer /, '') : null
```
Comment (flintinatux) [inline index.js]: The idiomatic way to do this in js is with the [case-insensitive `i` flag](https://devdocs.io/javascript/global_objects/regexp).  Also, Ramda has a [point-free `replace` function](https://devdocs.io/ramda/index#replace):

```js
const stripBearer =
  replace(/^Bearer /i, '')
```

The null-check is best left to the existing `enforce` function.  I'll make a separate comment explaining that.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@articulate/authentic",
-  "version": "0.1.2",
+  "version": "0.1.3",
```
Comment (flintinatux) [inline package.json]: Bumping the version isn't done until after the updates have been merged into `#master`.  After the PR is merged, a new version is created and published with the following flow:

```sh
> yarn version  # bumps the version and creates a matching git tag
> git push --tags origin master
> npm publish --access=public
```
Just move this back to `0.1.2` for now.

File: index.js [hunk 4]
```diff
@@ -50,11 +53,12 @@ const factory = opts => {
         .then(cacheClient(iss))
         .then(thrush(kid))
 
-  const verify = curryN(2, partialRight(promisify(jwt.verify), [ verifyOpts ]))
+  const verify = pipe(stripBearer, curryN(2, partialRight(promisify(jwt.verify), [ verifyOpts ])))
 
   const authentic = token =>
     Promise.resolve(token)
       .then(tapP(enforce))
+      .then(stripBearer)
```
Comment (flintinatux) [inline index.js]: The need to strip and null-check in multiple places is a 👃 that suggests we organize things a little better with some composition:

```js
// use unauthorized in enforce
const enforce = token =>
  token || Promise.reject(unauthorized('null token not allowed'))

const stripBearer = // <<-- move here to maintain alphebetic order
  replace(/^Bearer /i, '')

const unauthorized = err =>
  Promise.reject(Boom.wrap(err, 401))

const factory = opts => {
  // omiting other bits here...

  // remove stripBearer from verify
  const verify = curryN(2, partialRight(promisify(jwt.verify), [ verifyOpts ]))

  const authentic = token =>
    Promise.resolve(token)
      // remove tapP(enforce) here
      // remove stripBearer
      .then(decode)
      .then(tapP(checkIss))
      .then(getSigningKey)
      .then(chooseKey)
      .then(verify(token))
      .catch(unauthorized)

  // composeP to null-check and strip all in one place
  return composeP(authentic, stripBearer, tapP(enforce))
}
```

**Review by flintinatux — CHANGES_REQUESTED** [review]
See comments below.

**One general comment on workflow:**  please no more squashing and force-pushing.  Just make separate commits for your changes.  Force-pushing is problematic for others wishing to checkout locally and contribute.

When we merge it, we'll select the `Squash and Merge` option, which will rollup the whole PR into a single commit for us.  It'll keep the commit history nice and pretty.

**Comment by dpbackes** [comment]
@flintinatux updated. Rambda [doesn't support catches in composeP](https://github.com/ramda/ramda/issues/1799) so I just did a normal promise chain.

**Comment by flintinatux** [comment]
@dpbackes, that's why I suggested swapping the `new Error` in `enforce` with `unauthorized`, so that the error would be correctly formatted without catching, and the `.catch` could be left in the `authentic` function.

**Comment by flintinatux** [comment]
Hm, I'm noticing now that it should be `Boom.unauthorized` in `enforce`, since the custom `unauthorized` function wraps a native `Error` object.

**Comment by flintinatux** [comment]
squirrel

**Review by flintinatux — APPROVED** [review]

**Pull request merged**